cmd/binlog-collector/collector/collector.go

```diff
@@ -0,0 +1,184 @@
+	PXCUser        string
+	PXCPass        string
+	S3Endpoint     string
+	S3accessKeyID  string
+	S3accessKey    string
+	S3bucketName   string
+}
+
+func New(c Config) (*Controller, error) {
+	s3, err := storage.NewS3(c.S3Endpoint, c.S3accessKeyID, c.S3accessKey, c.S3bucketName, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+	lastSetFileName := "last-set.txt"
+	// get last binlog set stored on S3
+	lastSet, err := s3.GetObject(lastSetFileName)
+	if err != nil && !strings.Contains(err.Error(), "does not exist") {
+		return nil, errors.Wrap(err, "get lastset content")
+	}
+
+	return &Controller{
+		storage:         s3,
+		lastSet:         string(lastSet),
+		pxcUser:         c.PXCUser,
+		pxcPass:         c.PXCPass,
+		pxcServiceName:  c.PXCServiceName,
+		lastSetFileName: lastSetFileName,
+	}, nil
+}
+
+func (c *Controller) Run() error {
+	err := c.newDB()
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files:")
+	}
+
+	return nil
+}
+
+func (c *Controller) newDB() error {
+	host, err := c.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Controller) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Controller) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+c.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			return nodeArr[0] + "." + c.pxcServiceName, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (c *Controller) CollectBinLogs() error {
+	// get last uploaded binlog file name
+	binlogName, err := c.db.GetBinLogName(c.lastSet)
+	if err != nil {
+		return errors.Wrap(err, "get binlog name by set")
+	}
+	list, err := c.db.GetBinLogList()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+
+	upload := false
+	// if there are no uploaded files we going to upload every binlog file
+	if len(binlogName) == 0 {
+		upload = true
+	}
+
+	for _, binlog := range list {
+		if binlog == binlogName { // this check is for uploading starting from needed file
+			upload = true
+		}
+		if upload {
+			err = c.manageBinlog(binlog)
+			if err != nil {
+				return errors.Wrap(err, "manage binlog")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) manageBinlog(binlog string) error {
+	cmd := exec.Command("mysqlbinlog", "-R", "-h"+c.db.GetHost(), "-u"+c.pxcUser, "-p"+c.pxcPass, binlog)
```

Comment:
please set a password via `MYSQL_PWD` env variable.
reason - some people think that it is insecure to use a password on the command line.